File: 8-string-to-integer-atoi/8-string-to-integer-atoi.py
# Approach: Follow the Rules

# Time complexity: O(N) 
# N is the number of characters in the input string

# Space complexity: O(1)

class Solution:
    def myAtoi(self, s: str) -> int:
        sign = 1
        result = 0
        index = 0
        n = len(s)
        
        INT_MAX = pow(2, 31) - 1
        INT_MIN = -pow(2, 31)
        
        # discard all whitespaces from the beginning of the string
        while index < n and s[index] == ' ':
            index += 1
         
        # sign is +1 if positive or -1 if negative
        if index < n and s[index] == '+':
            sign = 1
            index += 1
        elif index < n and s[index] == '-':
            sign = -1
            index += 1
            
        # traverse next digits of the string and stop if it's not a digit
        while index < n and s[index].isdigit():
            digit = int(s[index])
            
            # check overflow and underflow conditions
            if ((result > INT_MAX // 10) or (result == INT_MAX // 10 and digit > INT_MAX % 10)):
                return INT_MAX if sign == 1 else INT_MIN
            
            # append current digit to the result
            result = 10 * result + digit
            index += 1
            
        return sign * result
         
        

File: 8-string-to-integer-atoi/8-string-to-integer-atoi.js
// Approach: Follow the Rules

// Time complexity: O(N) 
// N is the number of characters in the input string

// Space complexity: O(1)

/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function(s) {
    let res = 0;
    let sign = 1;
    let index = 0;
    let n = s.length;
    const INT_MAX = Math.pow(2, 31) - 1;
    const INT_MIN = -Math.pow(2, 31);
    
    while(index < n && s[index] === " ") {
        index++;
    }
    
    if(index < n && s[index] === "+") {
        sign = 1;
        index++;
    } else if (index < n && s[index] === "-") {
        sign = -1;
        index++;
    }
    
    while(index < n && !isNaN(s[index]) && s[index] !== " ") {
        let digit = s[index] - 0;
        if (res > Math.floor(INT_MAX / 10) || (res === Math.floor(INT_MAX / 10) && digit > INT_MAX % 10)) {
            return sign === 1 ? INT_MAX : INT_MIN
        }
        
        res = res * 10 + digit;
        index++;
    }
    
    return sign * res;
};